### PR TITLE
[PULL REQUEST] Add adaptive solver capability

### PR DESCRIPTION
**Pull request by jimmielin**
Hi GCST,

This patch contains the adaptive solver capability [(Lin et al., 2022 in review)](https://eartharxiv.org/repository/view/3468/) into GEOS-Chem 14.1.

The patch is separated into several commits:
* `0c0f493` builds the mechanism using KPP 3.0.0-rc.2, up from KPP 2.5.0 that GEOS-Chem uses. This brings in other changes from KPP apart from the `rosenbrock_autoreduce` solver. However, by default, the behavior of the `rosenbrock_autoreduce` solver is to be exactly like Rosenbrock. Zero differences are expected here.
* The other commits add the adaptive solver functionality:
    * binding to the `ICNTRL` and `RCNTRL` options in `fullchem_mod.F90`
    * adding `KppAutoReducerNVAR`, `KppcNONZERO`, and `KppAutoreduceThres` diagnostics for auto-reduce into a `KPPArDiags.` diagnostic collection
    * adding `KppTime` diagnostic for timing KPP integration time in each grid box into the `KppDiags.` collection
    * adding autoreduce_solver options in `geoschem_config.yml`:

```yaml
chemistry:
    autoreduce_solver:
      activate: false
      use_target_threshold:
        activate: true
        oh_tuning_factor: 5e-5
        no2_tuning_factor: 1e-4
      use_absolute_threshold:
        scale_by_pressure: true
        absolute_threshold: 100.0
      keep_halogens_active: false
      append_in_internal_timestep: false
```

These defaults represent the "standard configuration" of the adaptive solver as described in the manuscript (OH factor 5e-5, NO2 factor 1e-4, no keep halogens and no "append"), when `autoreduce_solver%activate` is set to `true`. Otherwise, with `false` as default, GEOS-Chem with this functionality will behave identically as before.

(Please feel free to squash the commits when merging)
 
This set of changes is expected to bring zero differences out-of-the-box and has been verified in the following way:
* Reference commit is `2b93644... Merge branch 'bugfix/set_zero_AOD_wavelength_for_FastJX' into dev`
* Ran `2b93644` against latest commit for 24 hours
* `$ ncdiff exp1_baseline_2/GEOSChem.SpeciesConc.20190701_0000z.nc4 exp0_unmodified/GEOSChem.SpeciesConc.20190701_0000z.nc4 exp1_exp0_diff.nc4`
* verified that `exp1_exp0_diff.nc4` contents are zero within numerical noise (approximately 1e-10 mol/mol)

This update should be merged together with https://github.com/geoschem/geos-chem/pull/1264 so that the molec cm-3 outputs can later be compared to measure the % difference auto-reduce solver will introduce, when enabled.

Thanks!
Haipeng

### CODE REVIEW

**Comment by yantosca** [comment]
Thanks  @jimmielin!  We'll prioritize this for 14.1.0.

**Comment by lizziel** [comment]
Hi @jimmielin, I am testing this out in GCHP and GEOS. It looks great in GCHP! I noticed that in your HISTORY.rc file for GC-Classic you added the diagnostics collection definition but did not add the name to the collections list at the top of the file. We can do that along with the run directory updates for other fullchem simulations, but I'm putting a note about it here so we don't forget.

**Comment by jimmielin** [comment]
Thanks Lizzie! Sorry I forgot about adding the collection name. Glad to hear it looks great in GCHP! I've been meaning to try it in GCHP as well to see if it relieves the load imbalance issues but haven't gotten to it yet.

File: KPP/fullchem/fullchem.kpp [hunk 1]
```diff
@@ -1,11 +1,12 @@
-#MINVERSION   2.5.0       { Need this version of KPP or later          }
-#INTEGRATOR   rosenbrock  { Use Rosenbrock integration method          }
-#LANGUAGE     Fortran90   { Generate solver code in Fortran90 ...      }
-#UPPERCASEF90 on          { ... with .F90 suffix (instead of .f90)     }
-#DRIVER       none        { Do not create gckpp_Main.F90               }
-#HESSIAN      off         { Do not create the Hessian matrix           }
-#MEX          off         { MEX is for Matlab, so skip it              }
-#STOICMAT     off         { Do not create stoichiometric matrix        }
+#MINVERSION   2.5.0                  { Need this version of KPP or later          }
```
Comment (yantosca) [inline KPP/fullchem/fullchem.kpp:1]: MINVERSION should be 2.6.0 or greater, as that is when the KPP adaptive solver came in.

File: run/GCClassic/geoschem_config.yml.templates/geoschem_config.yml.fullchem [hunk 2]
```diff
@@ -56,6 +56,17 @@ operations:
       activate: true
       use_static_bnd_cond: true
     gamma_HO2: 0.2
+    autoreduce_solver:
```
Comment (yantosca) [inline run/GCClassic/geoschem_config.yml.templates/geoschem_config.yml.fullchem:59]: Will also need to add this block to the geoschem_config.yml.Hg file.

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.fullchem [hunk 3]
```diff
@@ -393,6 +393,22 @@ COLLECTIONS: 'Restart',
                               'KppLuDecomps                  ',
                               'KppSubsts                     ',
                               'KppSmDecomps                  ',
+                              'KppTime                       ',
```
Comment (yantosca) [inline run/GCClassic/HISTORY.rc.templates/HISTORY.rc.fullchem:396]: Will also need to add this to the GCHP HISTORY.rc template files.

**Review by yantosca — APPROVED** [review]
This PR looks good.  It should be one of the last branches to be merged into dev/14.1.0 so that we can inline other changes to KPP mechanisms.

**Comment by lizziel** [comment]
@yantosca and @jimmielin, I made additional updates for the adaptive solver in GEOS, and pushed them to Bob's branch `feature/autoreduce_14.1`. Since that branch already has Haipeng's branch merged, I think we should we close this one out and make a PR using that branch instead. Alternatively, Haipeng could cherry-pick the 3 latest commits in that branch (2 from me, 1 from Bob).  How much work is left for the adaptive solver capability?

**Comment by lizziel** [comment]
Another option is create a draft PR with new branch, and specify in the conversation the required ordering of PRs that must precede it, which includes both this PR and geoschem/geos-chem PR #1264.

**Comment by lizziel** [comment]
@jimmielin and @msulprizio weighed in to close this PR and open a new one with the new branch. @yantosca, could you create a new PR from your branch?

**Comment by yantosca** [comment]
I can do that @lizziel @msulpzrizio @jimmielin.  Stay tuned.

**Comment by yantosca** [comment]
This PR is now superseded by #1482.  Closing.

**Pull request closed without merging**